File: Applied Course/4.Problem Solving/4.Problems on Stacks and Queues/2.Print Next Greater Element.c
#include<stdio.h>
#include<stdlib.h>
#define STACKSIZE 100
 
struct stack
{
    int top;
    int items[STACKSIZE];
};

void push(struct stack *ps, int x)
{
    if (ps->top == STACKSIZE-1)
    {
        printf("Error: stack overflown");
        exit(0);
    }
    else
    {
        ps->top += 1;
        int top = ps->top;
        ps->items [top] = x;
    }
}
 
int isEmpty(struct stack *ps)
{
    return (ps->top == -1)? 1 : 0;
}
 
int pop(struct stack *ps)
{
    int temp;
    if (ps->top == -1)
    {
        printf("Error: stack underflow n");
        exit(0);
    }
    else
    {
        int top = ps->top;
        temp = ps->items [top];
        ps->top -= 1;
        return temp;
    }
}
 

void nextGreaterElement(int *arr, int n)
{
    int i = 0;
    struct stack s;
    s.top = -1;  // initialize top with -1
    int element, next;
 
    /* push the first element to stack */
    push(&s, arr[0]);

    for (i=1; i<n; i++)
    {
        next = arr[i];
        if(!isEmpty(&s))
        {
            element = pop(&s); // pop the stack element, if it is not empty
 
            while(element < next)
            {
                printf("\n%d --> %d", element, next);   
                if(isEmpty(&s)) // if stack is empty break
                   break;
                element = pop(&s);
            }
            if (element > next)
                push(&s, element);
        }

        push(&s, next);
    }
 
    while(!isEmpty(&s))
    {
        element = pop(&s);
        next = -1;
        printf("\n%d --> %d", element, next);
    }
}
 

int main()
{
    int arr[]= {4, 5, 2, 25};
    int n = sizeof(arr)/sizeof(arr[0]);
    nextGreaterElement(arr, n);
    return 0;
}

File: Applied Course/4.Problem Solving/4.Problems on Stacks and Queues/2.Print Next Greater Element.py
import sys
class node: 

    def __init__(self, info): 
        self.info = info  
        self.next = None 


class Stack: 

    def __init__(self): 
        self.top = None
        
    
    def isEmpty(self):
        if self.top is None:
            return True
        return False
    
    def push(self,data):
        self.temp=node(data)
        if self.temp is None:
            print("Stack overflow")
            return
        self.temp.next=self.top
        self.top=self.temp
    
    def pop(self):
        if self.isEmpty():
            print("Stack Underflow")
            sys.exit(0)
        d=self.top.info
        self.top=self.top.next    
        return d
    
    def peek(self):
        if self.isEmpty():
            print("Stack Underflow")
            sys.exit(0)
        d=self.top.info
        return d
    def display(self):
        if self.isEmpty():
            print("Stack Underflow")
            sys.exit(0)
        self.p=self.top
        while self.p is not None:
            print(self.p.info)
            self.p=self.p.next
def nextGreaterEle(arr,n):
    s=Stack()
    s.push(arr[0])
    for i in range(1,n):
        temp=arr[i]
        if s.isEmpty() is not True:
            element=s.pop()
            while(element<temp):
                print("{} -> {}".format(element,temp))
                if s.isEmpty() is True:
                    break
                element=s.pop()
            if element>temp:
                s.push(element)
        s.push(temp)

    while s.isEmpty() is not True:
        element=s.pop()
        temp=-1
        print("{} -> {}".format(element,temp))
    
if __name__=='__main__':
    arr=[4, 3, 2, 1]
    n=len(arr)
    nextGreaterEle(arr,n)
